File: Module 02 Sort/main.go
package main

import (
	"fmt"
	"math/rand"
	"time"
)

func checkAvailable(arr []int) bool {
	seen := make(map[int]bool)
	for _, val := range arr {
		if val == 0 || seen[val] {
			return false
		}
		seen[val] = true
	}
	return true
}

func clone(original []int) []int {
	newArr := make([]int, len(original))
	copy(newArr, original)
	return newArr
}

func printArray(name string, arr []int) {
	fmt.Printf("%s: ", name)
	for _, val := range arr {
		fmt.Printf("%d ", val)
	}
	fmt.Println()
}

// Selection Sort
func SelectionSort(arr []int) []int {
	if !checkAvailable(arr) {
		panic("Invalid input")
	}
	out := []int{}
	working := clone(arr)
	for len(working) > 0 {
		minIdx := 0
		for i := 1; i < len(working); i++ {
			if working[i] < working[minIdx] {
				minIdx = i
			}
		}
		out = append(out, working[minIdx])
		working = append(working[:minIdx], working[minIdx+1:]...)
	}
	return out
}

// Insertion Sort
func InsertionSort(arr []int) []int {
	if !checkAvailable(arr) {
		panic("Invalid input")
	}
	out := clone(arr)
	for i := 1; i < len(out); i++ {
		key := out[i]
		j := i - 1
		for j >= 0 && out[j] > key {
			out[j+1] = out[j]
			j--
		}
		out[j+1] = key
	}
	return out
}

// Merge Sort
func MergeSort(arr []int) []int {
	if !checkAvailable(arr) {
		panic("Invalid input")
	}
	if len(arr) <= 1 {
		return arr
	}
	mid := len(arr) / 2
	left := MergeSort(arr[:mid])
	right := MergeSort(arr[mid:])
	return merge(left, right)
}

func merge(left, right []int) []int {
	result := []int{}
	i, j := 0, 0
	for i < len(left) && j < len(right) {
		if left[i] < right[j] {
			result = append(result, left[i])
			i++
		} else {
			result = append(result, right[j])
			j++
		}
	}
	result = append(result, left[i:]...)
	result = append(result, right[j:]...)
	return result
}

// Shell Sort
func ShellSort(arr []int) []int {
	if !checkAvailable(arr) {
		panic("Invalid input")
	}
	out := clone(arr)
	n := len(out)
	for gap := n / 2; gap > 0; gap /= 2 {
		for i := gap; i < n; i++ {
			temp := out[i]
			j := i
			for j >= gap && out[j-gap] > temp {
				out[j] = out[j-gap]
				j -= gap
			}
			out[j] = temp
		}
	}
	return out
}

// Quick Sort (custom pivot logic)
func QuickSort(arr []int) []int {
	if !checkAvailable(arr) {
		panic("Invalid input")
	}
	min, max := arr[0], arr[0]
	for _, v := range arr {
		if v < min {
			min = v
		}
		if v > max {
			max = v
		}
	}
	rand.Seed(time.Now().UnixNano())
	pivot := arr[rand.Intn(len(arr))]
	for pivot == min || pivot == max {
		pivot = arr[rand.Intn(len(arr))]
	}

	left := []int{}
	right := []int{}
	for _, val := range arr {
		if val < pivot {
			left = append(left, val)
		} else if val > pivot {
			right = append(right, val)
		}
	}

	leftSorted := InsertionSort(left)
	rightSorted := InsertionSort(right)

	result := append(leftSorted, pivot)
	result = append(result, rightSorted...)
	return result
}

func main() {
	original := []int{4, 9, 1, 5, 2}

	printArray("InsertionSort", InsertionSort(clone(original)))
	printArray("MergeSort", MergeSort(clone(original)))
	printArray("ShellSort", ShellSort(clone(original)))
	printArray("QuickSort", QuickSort(clone(original)))
	printArray("SelectionSort", SelectionSort(clone(original)))
}

File: Module 02 Sort/main.c
#include <stdio.h>
#include <stdlib.h>
#include <stdbool.h>
#include <time.h>
#include <string.h>

#define SIZE 5

// Check if input is valid: no duplicates, no zero
bool checkAvailable(int arr[], int size) {
    for (int i = 0; i < size; i++) {
        if (arr[i] == 0) return false;
        for (int j = i + 1; j < size; j++) {
            if (arr[i] == arr[j]) return false;
        }
    }
    return true;
}

// Print array
void printArray(int arr[], int size) {
    for (int i = 0; i < size; i++)
        printf("%d ", arr[i]);
    printf("\n");
}

// Selection Sort
void SelectionSort(int arr[], int size) {
    if (!checkAvailable(arr, size)) {
        printf("Invalid input.\n");
        return;
    }
    for (int i = 0; i < size - 1; i++) {
        int min_idx = i;
        for (int j = i + 1; j < size; j++) {
            if (arr[j] < arr[min_idx])
                min_idx = j;
        }
        int temp = arr[min_idx];
        arr[min_idx] = arr[i];
        arr[i] = temp;
    }
}

// Insertion Sort
void InsertionSort(int arr[], int size) {
    if (!checkAvailable(arr, size)) {
        printf("Invalid input.\n");
        return;
    }
    for (int i = 1; i < size; i++) {
        int key = arr[i];
        int j = i - 1;
        while (j >= 0 && arr[j] > key) {
            arr[j + 1] = arr[j];
            j--;
        }
        arr[j + 1] = key;
    }
}

// Merge Sort
void merge(int arr[], int l, int m, int r) {
    int n1 = m - l + 1, n2 = r - m;
    int L[n1], R[n2];
    for (int i = 0; i < n1; i++) L[i] = arr[l + i];
    for (int j = 0; j < n2; j++) R[j] = arr[m + 1 + j];
    int i = 0, j = 0, k = l;
    while (i < n1 && j < n2)
        arr[k++] = (L[i] <= R[j]) ? L[i++] : R[j++];
    while (i < n1) arr[k++] = L[i++];
    while (j < n2) arr[k++] = R[j++];
}

void MergeSort(int arr[], int l, int r) {
    if (!checkAvailable(arr + l, r - l + 1)) {
        printf("Invalid input.\n");
        return;
    }
    if (l < r) {
        int m = l + (r - l) / 2;
        MergeSort(arr, l, m);
        MergeSort(arr, m + 1, r);
        merge(arr, l, m, r);
    }
}

// Shell Sort
void ShellSort(int arr[], int size) {
    if (!checkAvailable(arr, size)) {
        printf("Invalid input.\n");
        return;
    }
    for (int gap = size / 2; gap > 0; gap /= 2) {
        for (int i = gap; i < size; i++) {
            int temp = arr[i], j;
            for (j = i; j >= gap && arr[j - gap] > temp; j -= gap)
                arr[j] = arr[j - gap];
            arr[j] = temp;
        }
    }
}

// QuickSort (custom version, not recursive)
void QuickSort(int arr[], int size) {
    if (!checkAvailable(arr, size)) {
        printf("Invalid input.\n");
        return;
    }

    int max = arr[0], min = arr[0];
    for (int i = 1; i < size; i++) {
        if (arr[i] > max) max = arr[i];
        if (arr[i] < min) min = arr[i];
    }

    srand(time(NULL));
    int pivot;
    do {
        pivot = arr[rand() % size];
    } while (pivot == min || pivot == max);

    int left[SIZE], right[SIZE];
    int l = 0, r = 0;
    for (int i = 0; i < size; i++) {
        if (arr[i] < pivot) left[l++] = arr[i];
        else if (arr[i] > pivot) right[r++] = arr[i];
    }

    // Sort left and right manually
    SelectionSort(left, l);
    SelectionSort(right, r);

    int idx = 0;
    for (int i = 0; i < l; i++) arr[idx++] = left[i];
    arr[idx++] = pivot;
    for (int i = 0; i < r; i++) arr[idx++] = right[i];
}

int main() {
    int original[SIZE] = {4, 9, 1, 5, 2};
    int a[SIZE];

    // Insertion Sort
    memcpy(a, original, sizeof(a));
    InsertionSort(a, SIZE);
    printf("InsertionSort: "); printArray(a, SIZE);

    // Merge Sort
    memcpy(a, original, sizeof(a));
    MergeSort(a, 0, SIZE - 1);
    printf("MergeSort: "); printArray(a, SIZE);

    // Shell Sort
    memcpy(a, original, sizeof(a));
    ShellSort(a, SIZE);
    printf("ShellSort: "); printArray(a, SIZE);

    // QuickSort
    memcpy(a, original, sizeof(a));
    QuickSort(a, SIZE);
    printf("QuickSort: "); printArray(a, SIZE);

    // Selection Sort
    memcpy(a, original, sizeof(a));
    SelectionSort(a, SIZE);
    printf("SelectionSort: "); printArray(a, SIZE);

    return 0;
}